### ISSUE TITLE
[inductor] unexpected cuda:0 device usage when compiling and runing a model on cuda:1

### ISSUE THREAD

**Issue by vfdev-5** [pytorch/pytorch#124854]
When user works with "cuda:1" device and compiles a model, there is unexpected cuda context initialization for device "cuda:0" happenning, which can be surprising to the user seeing with nvidia-smi the device 0 utilisation.

**Reproduction code:**
```python
import torch
from torchvision.models import resnet18

def print_memory_usage():
    for d in [0, 1]:
        stats = torch.cuda.memory_stats(device=d)
        m = stats["allocated_bytes.all.allocated"] + stats["inactive_split_bytes.all.allocated"] + stats["reserved_bytes.all.allocated"]
        print(f"\t- CUDA Device: {d}, allocated + reserved + non-released in MB: {m / 1024 / 1024}")

device = "cuda:1"
model = resnet18()
compiled_model = torch.compile(model)

print("--- Before compiled model to device")
print_memory_usage()

compiled_model.to(device)
x = torch.rand(16, 3, 320, 320, device=device)

print("--- Before compiled model forward")
print_memory_usage()

y = compiled_model(x)

print("--- Before compiled model backward")
print_memory_usage()

y.sum().backward()

print("--- After compiled model backward")
print_memory_usage()
```
Output:
```
--- Before compiled model to device
        - CUDA Device: 0, allocated + reserved + non-released in MB: 0.0
        - CUDA Device: 1, allocated + reserved + non-released in MB: 0.0
--- Before compiled model forward
        - CUDA Device: 0, allocated + reserved + non-released in MB: 0.0
        - CUDA Device: 1, allocated + reserved + non-released in MB: 192.966796875
--- Before compiled model backward
        - CUDA Device: 0, allocated + reserved + non-released in MB: 8.044921875    # <--- this should be zero
        - CUDA Device: 1, allocated + reserved + non-released in MB: 2054.27197265625
--- After compiled model backward
        - CUDA Device: 0, allocated + reserved + non-released in MB: 8.044921875    # <--- this should be zero
        - CUDA Device: 1, allocated + reserved + non-released in MB: 5654.61962890625
```

Similar, unexpected cuda device utilisation can be seen when we compile a model and run on CPU inputs.

All of them are mostly due to pattern registrations using cuda device if a cuda is available (and no matter the input device).

In [this PR](https://github.com/pytorch/pytorch/pull/124722), I tried to fix this problem but this ends up into the pattern registration device choice, for example:
https://github.com/pytorch/pytorch/blob/bf834d388b9037a27675f3d27a008b323bec4311/torch/_inductor/fx_passes/pad_mm.py#L444-L448

an open question whether we would like to:
1) register patterns multiple times according to the input's device and add `exist_ok` option for already registered device agnostic patterns. 

2) or register patterns once (current behaviour) and use current cuda device from the inputs. This option will still howver show cuda:0 utilization when first compiled model call is done on cpu input...

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @muchulee8 @ColinPeppler @amjames @desertfire @bdhirsh @anijain2305 @peterbell10

**Comment by IvanKobzarev** [comment]
Doing, issue scrubbing, still open. Repro ready for copy paste:
```
import torch
from torchvision.models import resnet18

def print_memory_usage():
    for d in ["cuda:0", "cuda:1"]:
        stats = torch.cuda.memory_stats(device=d)
        m = stats["allocated_bytes.all.allocated"] + stats["inactive_split_bytes.all.allocated"] + stats["reserved_bytes.all.allocated"]
        print(f"\t- CUDA Device: {d}, allocated + reserved + non-released in MB: {m / 1024 / 1024}")

device = "cuda:1"
model = resnet18()
compiled_model = torch.compile(model)

print("--- Before compiled model to device")
print_memory_usage()

compiled_model.to(device)
x = torch.rand(16, 3, 320, 320, device=device)

print("--- Before compiled model forward")
print_memory_usage()

y = compiled_model(x)

print("--- Before compiled model backward")
print_memory_usage()

y.sum().backward()

print("--- After compiled model backward")
print_memory_usage()
```
```
--- Before compiled model to device
        - CUDA Device: cuda:0, allocated + reserved + non-released in MB: 0.0
        - CUDA Device: cuda:1, allocated + reserved + non-released in MB: 0.0
--- Before compiled model forward
        - CUDA Device: cuda:0, allocated + reserved + non-released in MB: 0.0
        - CUDA Device: cuda:1, allocated + reserved + non-released in MB: 192.966796875
/data/users/ivankobzarev/a/pytorch/torch/_inductor/compile_fx.py:170: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
  warnings.warn(
--- Before compiled model backward
        - CUDA Device: cuda:0, allocated + reserved + non-released in MB: 8.044921875
        - CUDA Device: cuda:1, allocated + reserved + non-released in MB: 43096.06298828125
--- After compiled model backward
        - CUDA Device: cuda:0, allocated + reserved + non-released in MB: 8.044921875
        - CUDA Device: cuda:1, allocated + reserved + non-released in MB: 248933.08740234375
```

**Comment by eellison** [comment]
This repros even with aot_eager:

```
import torch
from torchvision.models import resnet18

def print_memory_usage():
    for d in ["cuda:0", "cuda:1"]:
        stats = torch.cuda.memory_stats(device=d)
        if not stats:
            print(f"\t- CUDA Device: {d}, no memory stats")
            continue
        m = stats["allocated_bytes.all.allocated"] + stats["inactive_split_bytes.all.allocated"] + stats["reserved_bytes.all.allocated"]
        print(f"\t- CUDA Device: {d}, allocated + reserved + non-released in MB: {m / 1024 / 1024}")

device = "cuda:1"
model = resnet18()
compiled_model = torch.compile(model, backend="aot_eager")

print("--- Before compiled model to device")
print_memory_usage()

compiled_model.to(device)
x = torch.rand(16, 3, 320, 320, device=device)

print("--- Before compiled model forward")
print_memory_usage()

y = compiled_model(x)

print("--- Before compiled model backward")
print_memory_usage()

y.sum().backward()

print("--- After compiled model backward")
print_memory_usage()
```


--- Before compiled model backward
        - CUDA Device: cuda:0, allocated + reserved + non-released in MB: 4.0
        - CUDA Device: cuda:1, allocated + reserved + non-released in MB: 3252.40673828125
